### ISSUE TITLE
OAS3 Links Support & Rewrite Links

### ISSUE THREAD

**Issue by philsturgeon** [stoplightio/prism#238]
One request came in for Prism v2 and help with [replacing hosts in JSON:API links](https://community.stoplight.io/t/mocking-replace-url-in-responses-hateoas/567), and I was wondering if we could solve that in Prism v3.

So, taking their link-in-the-body example:

![image](https://user-images.githubusercontent.com/67381/55742467-0c221100-5a30-11e9-94a5-bd77314ae453.png)

It is a little unfortunate they have had to pollute their spec with a variable, which will show up and look funny in other tools. Instead, what if they hardcode all links to match the content of a valid server. 

So, in their spec, define the servers array:

```
servers: 
  - url: https://api.foo.com
    description: Production server

  - url: https://staging-api.foo.com
    description: Staging server
```

Continuging with their example: 

```js
"links": {
  "self": "https://api.foo.com/api/shared_file_categories"
}
```

This is a nice default because it includes the production server, and it is only a problem when Prism returns that link. When clients follow it, they end up on production, so boo, let's turn that into a local URL.

If we add a CLI option `--rewrite-urls`, it could be implemented as a Fastly middleware which looks through the JSON/XML/HTML content for a URL which matches a known servers.url + a valid path as defined in the spec, it rewrites the server URL portion (everything before the path) of that link to whatever prism is currently running under. 

If Prism was running as usual:

```
$ prism mock links.yaml --rewrite-urls
http://127.0.0.1:4010
```

...and the calls were as usual:

```
http GET http://127.0.0.1:4010
```

The links would be rewritten to:

```js
"links": {
  "self": "http://127.0.0.1:4010/api/shared_file_categories"
}
```

If Prism was called in conjunction with server validation (missing from Prism v3 currently but implemented in v2):

```
http GET http://127.0.0.1:4010?__server=https://staging-api.foo.com
```

then we would see links rewritten like this:

```js
"links": {
  "self": "https://staging-api.foo.com/api/shared_file_categories"
}
```

What do you think?

I started off considering this somehow related to [OpenAPI "Links"](https://swagger.io/docs/specification/links/) but I think it can be done without going anywhere near them, as it is links in the body we are struggling with currently.

**Comment by pytlesk4** [comment]
@philsturgeon interesting idea, and I like it. 

The main concerns I have with `--rewrite-urls` as a cli option, we open pandoras box, and we start to specialize prism towards certain api designs.

Maybe instead of a cli option, it could be a plugin instead? I know the new version of prism doesn't have a plugin system just yet, but would be pretty cool. We could provide some of our own plugins initially, and if users wanted to, they could write their own.

**Comment by philsturgeon** [comment]
I don't really mind what code the CLI is enabling. When proposing this I figured the CLI option would enable the middleware, the middleware then interacts with the JS library, etc.

**Comment by pytlesk4** [comment]
Gotcha, what about something like `--plugin @stoplight/rewrite-urls`, then it doesn't look like something we officially support(unless of course that is our plugin). And you could pass in as many plugins as you want, `--plugin @stoplight/rewrite-urls --plugin @stoplight/cors` or `--plugin @stoplight/rewrite-urls,@stoplight/cors`. 

Just an idea that would need to be fleshed out, couple concerns from a user perspective is what order do plugins run in? I have some ideas around that. Adding plugins would mean we need to design a plugin interface, document, ect so that users could build them.

**Comment by philsturgeon** [comment]
I think we can wrap up official plugins with a bit of a nicer API than that, right? Save a lot of faff? 

To me, common functionality can be in the core, and as things get more outlandish they can go elsewhere, but this is a pretty common thing. Building a REST API is pretty much our whole thing, and supporting links is necessary for that.

**Comment by marbemac** [comment]
Neat idea! Plugins are cool but a HUGE topic.

Thinking out loud - for this, would a more generic "rewrite/replace" type arg get the job done? Something like:

```bash
prism mock links.yaml --rewrite "https://api.foo.com" "http://127.0.0.1:4010"
```

**Comment by distinctgrey** [comment]
I agree with @marbemac, plugins would be awesome but there's a lot more to that than simply creating a plugin architecture.

What I don't understand is why this should either be a plugin or new CLI parameter, as Stoplight already offers a system for extending mocking behavior through Scenarios?
Although a bit clunky and rather buggy, it _does_ more or less work.

Just an idea: maybe these prism "plugins" could take the form of Stoplight-provided Scenario steps one can use as building blocks?

**Comment by philsturgeon** [comment]
@distinctgrey The generic functionality provided can be exposed in multiple ways. I work mainly with the CLI tools so that’s the example I’m using, but for it to exist it will need work done on the core libraries. With that core work done it will be possible for us to expose that functionality through the various user interfaces of the Stoplight platform. 

@marbemac one downside with that —rewrite approach is that maybe it’s a little clunky to type? Generic things are good, but I feel like this is so generic it’s losing some of its sheen. It will be something we have to document as a use case, like “if you want to support HATEOAS links in the body, here is a little trick” instead of just a switch you enable and the use case happens for you through the existing conventions in Prism. 

I feel like folks might ask for generic rewrites in the future, and we can add them as a “do whatever” feature, but perhaps we can have both for the sake of the users fingers. :)
